accountserver/replicator.go

```diff
@@ -650,6 +719,134 @@ func (r *Replicator) Run() {
+	// TODO: add something here that will delete the container if it is empty
+	// the reaper will just make 2 passes to get all the way done
+	marker := ""
+	objs, resp := dc.GetContainer(cont, marker, "", 10000, "", "", false, map[string]string{})
+	var obj *nectar.ObjectRecord
+	hdrs := map[string]string{}
+	for len(objs) > 0 {
+		obj, objs = objs[0], objs[1:]
+		contObjChan <- &contObj{cont, obj.Name}
+		marker = obj.Name
+		if len(objs) == 0 {
+			objs, resp = dc.GetContainer(cont, marker, "", 10000, "", "", false, map[string]string{})
+		}
+		select {
+		case <-canceler:
+			return nil
+		default:
+		}
+	}
+	hdrs["X-Timestamp"] = common.GetTimestamp()
+	if resp = dc.DeleteContainer(cont, hdrs); resp == nil || resp.StatusCode/100 != 2 {
+		r.logger.Debug("invalid reap cont resp", zap.String("container", cont))
+	}
+	return nil
+}
+
+type contObj struct {
+	cont string
+	obj  string
+}
+
+func (r *Replicator) reapAccount(dbFile string, canceler chan struct{}) error {
+	db, err := sqliteOpenAccount(dbFile)
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+	if d, err := db.IsDeleted(); err != nil {
+		return err
+	} else if !d {
+		return fmt.Errorf("Trying to delete an active db: %s", dbFile)
+	}
+	info, err := db.GetInfo()
+	if err != nil {
+		return err
+	}
+	dc, err := client.NewDirectClient(info.Account, srv.DefaultConfigLoader{}, r.certFile, r.keyFile, r.logger)
+	if err != nil {
+		r.logger.Error("Could not create client to reap account.", zap.String("account", info.Account), zap.Error(err))
+	}
+	wg := sync.WaitGroup{}
+	conc := 20 // TODO: make config
+	wg.Add(conc)
+	contObjChan := make(chan *contObj)
+	var objsDeleted int64
+
+	for i := 0; i < conc; i++ {
+		go func() {
+			defer wg.Done()
+			for {
+				co := <-contObjChan
+				if co == nil {
+					return
+				}
+				if resp := dc.DeleteObject(co.cont, co.obj, map[string]string{"X-Timestamp": common.GetTimestamp()}); resp == nil || resp.StatusCode/100 != 2 {
+					r.logger.Debug("invalid reap object resp", zap.String("container", co.cont), zap.String("obj", co.obj))
+				} else {
+					atomic.AddInt64(&objsDeleted, 1)
+				}
+				r.reaperCheckin <- struct{}{}
+			}
+		}()
+	}
+	marker := ""
+	conts, err := db.ListContainers(1000, marker, "", "", "", false)
+	if err != nil {
+		return err
+	}
+	var contr interface{}
+	for len(conts) > 0 {
+		contr, conts = conts[0], conts[1:]
+		cont, ok := contr.(*ContainerListingRecord)
+		if ok {
+			if err = r.reapContainer(cont.Name, dc, contObjChan, canceler); err != nil {
+				r.logger.Error("error reaping container", zap.String("account", info.Account), zap.String("container", cont.Name), zap.Error(err))
+			}
+			marker = cont.Name
+		} else {
+			r.logger.Error("invalid listing", zap.String("record", fmt.Sprintf("%v", contr)))
+			break
+		}
+		if len(conts) == 0 {
+			conts, err = db.ListContainers(1000, marker, "", "", "", false)
+			if err != nil {
+				return err
```

Comment:
Seems like this would leave all the background stuff running. Should `break` then return the err below?